### PR TITLE
Store: Kitchen sink SQL implementation (do not merge)

### PR DESCRIPTION
**Pull request by ryantxu**
This PR is my best attempt at "all the things" working together at the same time.  As parts of it get mergable, we have been splitting them into their own PRs to make progress.

-----

This PR adapts the `file` table and the `dashboards_*` tables to create a structure that _should_ support a wide range of features.

Now that we do not support MySQL 5.6 we do not need special indexing in case the key is longer than 191 🎉 

This uses a "key" as the primary key -- this is essentially a file path that makes things globally unique.  It will include the orgId and whatever other structures we think make sense to organize the objects within the store

Questions:
* modified vs updated?  current tables use "updated" (same length as created!) but API currently uses "modified" 🤷 

TODO:
- [x] Read
- [x] Write
- [ ] Query

Observations:
* UserInfo object is awkward... can we use a string?
* GRN vs Kind + UID... also awkward
* Likely need explicit query response object with different "things" than exist in RawObject

### CODE REVIEW

File: pkg/services/store/object/sqlstash/sql_storage_server.go [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	modifier := userFromContext(ctx)
+	key := fmt.Sprintf("%d/%s.%s", modifier.OrgID, r.UID, r.Kind)
+
+	rsp := &object.WriteObjectResponse{}
+
+	err = s.sess.WithTransaction(ctx, func(tx *session.SessionTx) error {
```
Comment (ryantxu) [inline pkg/services/store/object/sqlstash/sql_storage_server.go:86]: @zserge -- any red flags with this approach?

I am still trying to learn how to best use sqlx here :)

File: pkg/services/sqlstore/session/session.go [hunk 2]
```diff
@@ -30,6 +30,10 @@ func (gs *SessionDB) Select(ctx context.Context, dest interface{}, query string,
 	return gs.sqlxdb.SelectContext(ctx, dest, gs.sqlxdb.Rebind(query), args...)
 }
 
+func (gs *SessionDB) Query(ctx context.Context, query string, args ...interface{}) (*sql.Rows, error) {
+	return gs.sqlxdb.QueryContext(ctx, gs.sqlxdb.Rebind(query), args...)
```
Comment (ryantxu) [inline pkg/services/sqlstore/session/session.go:34]: @zserge -- any reason not to use the query structures?

File: packages/grafana-ui/src/components/TimeSeries/utils.ts [hunk 3]
```diff
@@ -223,7 +223,14 @@ export const preparePlotConfigBuilder: UPlotConfigPrepFn<{
...
           centeredZero: customConfig.axisCenteredZero,
-          range: customConfig.stacking?.mode === StackingMode.Percent ? [0, 1] : undefined,
+          range:
+            customConfig.stacking?.mode === StackingMode.Percent
+              ? (u: uPlot, dataMin: number, dataMax: number) => {
+                  dataMin = dataMin < 0 ? -1 : 0;
+                  dataMax = dataMax > 0 ? 1 : 0;
+                  return [dataMin, dataMax];
+                }
+              : undefined,
```
Comment (torkelo) [inline packages/grafana-ui/src/components/TimeSeries/utils.ts:233]: unrelated change?

File: pkg/services/sqlstore/migrations/object_store_mig.go [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+			// The raw object body (any byte array)
+			{Name: "body", Type: migrator.DB_Blob, Nullable: false},
+			{Name: "size", Type: migrator.DB_BigInt, Nullable: false},
+			{Name: "etag", Type: migrator.DB_NVarchar, Length: 32, Nullable: false}, // md5(body)
+			{Name: "version", Type: migrator.DB_NVarchar, Length: 128, Nullable: false},
+
+			// Who changed what when
+			{Name: "updated", Type: migrator.DB_DateTime, Nullable: false},
+			{Name: "created", Type: migrator.DB_DateTime, Nullable: false},
+			{Name: "updated_by", Type: migrator.DB_Int, Nullable: false}, // joined to user table
```
[the remark this answers is not in the document]
Comment (bergquist) [reply pkg/services/sqlstore/migrations/object_store_mig.go:33]: We did decide to use the sql database instead of files but I'd like to keep the file option open as long as possible.

File: pkg/services/store/object/http.go [hunk 5]
```diff
@@ -47,6 +51,9 @@ func (s *httpObjectStore) RegisterHTTPRoutes(route routing.RouteRegister) {
 	route.Get("/history/*", reqGrafanaAdmin, routing.Wrap(s.doGetHistory))
 	route.Get("/list/*", reqGrafanaAdmin, routing.Wrap(s.doListFolder)) // Simplified version of search -- path is prefix
 	route.Get("/search", reqGrafanaAdmin, routing.Wrap(s.doSearch))
+
+	// File upload
+	route.Post("/upload", reqGrafanaAdmin, routing.Wrap(s.doUpload))
```
Comment (bergquist) [inline pkg/services/store/object/http.go:56]: I'm quite concerned about starting to use the object store for any kind of file right now. can we defer this?

Uploading the files to blog storage and keeping a reference of them in the object store would probably be my suggestion.

Comment (ryantxu) [reply pkg/services/store/object/http.go:56]: 🤣 lots of catch 22s here!!!!  What about the existing `file` sql store?!  that is pretty much identical to this table (without the summary info that gets saved in the same transaction)

I agree blob storage is the best path -- but I think that comes later, after we agree on an object store interface that is comfortable with "objects"

For sure we need configurable rate/quota limiting.  But hard to say uploading images+or csv is worse than 20mb dashboards 🤷

File: pkg/services/sqlstore/migrations/object_store_mig.go [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+	objectTable := migrator.Table{
+		Name: "object",
+		Columns: []*migrator.Column{
+			// Object key contains everything required to make it unique across all instances
+			// orgId+kind+uid+scope?
+			{Name: "key", Type: migrator.DB_NVarchar, Length: 1024, Nullable: false, IsPrimaryKey: true},
+
+			// If objects are organized into folders (ie '/' exists in the uid),
+			// this will optimize the common use case of listing all files in a folder
+			{Name: "parent_folder_key", Type: migrator.DB_NVarchar, Length: 1024, Nullable: false},
```
[the remark this answers is not in the document]
Comment (ryantxu) [reply pkg/services/sqlstore/migrations/object_store_mig.go:19]: For sure the goal is to get this structured so it has fast queries and that tenant_id is the root.  My real target with "key" is GCS/S3 etc -- walking up+down the tree should give us good isolation.  If we want to implement additional folder+security metadata on top of that we can:

Current state in this PR:
https://github.com/grafana/grafana/pull/56348/files#diff-ce50f456e6212a305e7bf52b8520494e742963a5123a1ac5987c2f7afd3b18a2R29

File: pkg/services/sqlstore/migrations/object_store_mig.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (ryantxu) [reply pkg/services/sqlstore/migrations/object_store_mig.go:33]: Absolutely -- we need to make sure this table is never joined with anything outside of it.  The user lookups will happen outside of the store.

Any thoughts on making the user references strings?  They will need to be passed to another service that can add additional display information.

File: pkg/services/sqlstore/migrations/object_store_mig.go [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+
+			// For objects that are synchronized from an external source (ie provisioning or git)
+			{Name: "sync_src", Type: migrator.DB_Text, Nullable: true},
+			{Name: "sync_time", Type: migrator.DB_DateTime, Nullable: true},
+
+			// Summary data (always extracted from the `body` column)
+			{Name: "name", Type: migrator.DB_NVarchar, Length: 255, Nullable: false},
+			{Name: "description", Type: migrator.DB_NVarchar, Length: 255, Nullable: true},
+			{Name: "labels", Type: migrator.DB_Text, Nullable: true}, // JSON object
+			{Name: "fields", Type: migrator.DB_Text, Nullable: true}, // JSON object
```
[the remark this answers is not in the document]
Comment (ryantxu) [reply pkg/services/sqlstore/migrations/object_store_mig.go:44]: Each "kind" can return a set of fields that can depend on the body -- these can be returned quickly in the query results.  Still TBD if we should support searching over them

https://github.com/grafana/grafana/blob/main/pkg/services/store/kind/png/summary_test.go#L28

File: pkg/services/store/object/http.go (hunk 5 above)
Comment (torkelo) [reply pkg/services/store/object/http.go:56]: We have 20mb dashboard json blobs ? How is that possible? :)

Comment (ryantxu) [reply pkg/services/store/object/http.go:56]: panels with embedded data (testdata etc) / snapshots

When rolling out search in cloud -- the spikes we would see had more to do with large dashboards than with number of dashboards.

File: pkg/services/sqlstore/migrations/object_store_mig.go [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+
+			// Commit message
+			{Name: "message", Type: migrator.DB_Text, Nullable: false}, // defaults to empty string
+		},
+		Indices: []*migrator.Index{
+			{Cols: []string{"key", "version"}, Type: migrator.UniqueIndex},
+		},
+	}
+
+	// Define access based on prefix rules (eg... .htaccess (- ‿◦ ))
```
[the remark this answers is not in the document]
Comment (ryantxu) [reply pkg/services/sqlstore/migrations/object_store_mig.go:113]: this is still exploratory, and will take some work to know if it makes sense to continue.  For sure authn should be managed by a separate service, but I am not yet convinced authz should be -- this will matter more if we choose to use path based authorization.  The nice part of this is that it would be more self contained and have all the same history and audit controls.

That said -- I am far from convinced this is the right path forward, but it is much simpler and easier to embed/control than our current approach
